// ngx-panzoom/esm2020/lib/panzoom.module.mjs
import { NgModule } from '@angular/core';
import { PanZoomComponent } from './panzoom.component';
import * as i0 from "@angular/core";
export class NgxPanZoomModule {
}
NgxPanZoomModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.0", ngImport: i0, type: NgxPanZoomModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxPanZoomModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.0", ngImport: i0, type: NgxPanZoomModule, declarations: [PanZoomComponent], exports: [PanZoomComponent] });
NgxPanZoomModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.0", ngImport: i0, type: NgxPanZoomModule, providers: [], imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.0", ngImport: i0, type: NgxPanZoomModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [],
                    declarations: [PanZoomComponent],
                    providers: [],
                    exports: [PanZoomComponent]
                }]
        }] });
